#!/bin/sh

# This shell script creates a list of all differentially expressed genes (DEG) that are listed in the input file(s). Having a master list like this can be helpful when working with multiple comparisons with data spread over multiple files.

# It takes as input the 'DE results - *.csv' files generated by the Nanostring nSolver program. This file format contains:
#           column 1 - gene symbols with an '-mRNA' suffix
#           column 2 - the log2 fold change (log2FC) of the
#                      normalised transcript count in the data
#                      sets being compared
#           column 10 - the adjusted p value

# Note that for the script to correctly identify DEG the log2FC and p values must be in decimal notation and not scientific notation (eg. 0.0000739 not 7.39E-05). The csv files produced by nSolver contain a mix of decimal and scientific values. It is necessary to first convert them to all to decimals. As a workaround I've been doing this using Excel, formatting the cells to the general number format. Not ideal I know but works for now.

# step 1 - save files to be read from in their own directory, and set that as the current working directory
# step 2 - edit the THRESH and PVAL values defined in the code below to the desired thresholds for calling DEG (the default here is a log2FC of >= 1 or <= -1, and p value <= 0.05).
# step 3 - run the code. Output will be file 'All_DEG_list.txt' that lists all identified DEG gene symbols without the '-mRNA' suffix.

THRESH=1
PVAL=0.05

LOWER_THRESH=`expr $THRESH \* -1`

cat *.csv | awk -F "\"*,\"*" '{ if(($2 >= "'"$THRESH"'" || $2 <= "'"$LOWER_THRESH"'") && ( $10 <= "'"$PVAL"'")) print $1 }' | sed 's/-mRNA//' | sort | uniq > All_DEG_list.txt
